### ISSUE TITLE
Error running ESC on IntMathOps example

### ISSUE THREAD

**Issue by ducmle** [OpenJML/OpenJML#580]
Dear David, 

Following on from #574, I am currently having another problem running OpenJML (0.8.24) with Z3_4_3 on the IntMathOps example in the Leavens paper (titled: [Preliminary Design of JML:
A Behavioral Interface Specification Language for Java](https://scholar.google.com.vn/scholar?cluster=1389285402619830442&hl=en&as_sdt=2005&sciodt=0,5&as_vis=1))

The command to run and the errors are given below. The IntMathOps class is attached for your reference. It seems that the prover cannot interprete the pre-condition assertion.

Thank you very much for your help.

> bin/openjml esc examples/src/jmleg/IntMathOps.java
>
examples/src/jmleg/IntMathOps.java:20: warning: The prover cannot establish an assertion (Precondition: openjml/openjml.jar(specs18/java/lang/Math.jml):263: ) in method isqrt
    return (int) Math.sqrt(y);
                          ^
openjml/openjml.jar(specs18/java/lang/Math.jml):263: warning: Associated declaration: examples/src/jmleg/IntMathOps.java:20: 
    public static /*@ pure @*/ double sqrt(double a);
                                      ^
openjml/openjml.jar(specs18/java/lang/Math.jml):247: warning: Precondition conjunct is false: Double.isNaN(a) || a < 0.0
      @   requires Double.isNaN(a) || a < 0.0;
                   ^
openjml/openjml.jar(specs18/java/lang/Math.jml):250: warning: Precondition conjunct is false: a == Double.POSITIVE_INFINITY
      @   requires a == Double.POSITIVE_INFINITY;
                   ^
openjml/openjml.jar(specs18/java/lang/Math.jml):253: warning: Precondition conjunct is false: isPositiveZero(a)
      @   requires isPositiveZero(a);
                                 ^
openjml/openjml.jar(specs18/java/lang/Math.jml):256: warning: Precondition conjunct is false: isNegativeZero(a)
      @   requires isNegativeZero(a);
                                 ^
openjml/openjml.jar(specs18/java/lang/Math.jml):259: warning: Precondition conjunct is false: a < Double.POSITIVE_INFINITY
      @	  requires a > 0.0 && a < Double.POSITIVE_INFINITY;
       	                      ^
examples/src/jmleg/IntMathOps.java:20: warning: The prover cannot establish an assertion (Postcondition: examples/src/jmleg/IntMathOps.java:14: ) in method isqrt
    return (int) Math.sqrt(y);
    ^
examples/src/jmleg/IntMathOps.java:14: warning: Associated declaration: examples/src/jmleg/IntMathOps.java:20: 
     @ ensures 0.0 <= \result
       ^

[IntMathOps.java.txt](https://github.com/OpenJML/OpenJML/files/1385101/IntMathOps.java.txt)

**Comment by davidcok** [comment]
The problem here is that OpenJML does not yet have sufficient knowledge of real arithmetic, even the properties of an int cast to double.
I’ll keep this example as a test case.

- David

**Comment by davidcok** [comment]
Note also that provers as quite slow at real arithmetic.
If you are just looking for instructional examples, stick with integer arithmetic.

**Comment by ducmle** [comment]
Dear David, 

Ok, I will keep this in mind. By the way, is there a comprehensive list of the current ESC capabilities of OpenJML (perhaps w.r.t each type of solver?) so that we can know what to expect when using this functionality? The user guide seems only partially written and a bit lacking (?)

The reason I ask this question is because I am using OpenJML to design and code a relatively simple OOP example (for study purpose of course.) A key part of this exercise is to demonstrate that OpenJML can detect implementation errors in the code. I will report issues as I come across them.

Thanks,
Duc

**Comment by davidcok** [comment]
I agree the user guide is lacking. It and a new reference manual is on my list of tasks, but are taking second place to the project I have between now and the end of November.

- David
